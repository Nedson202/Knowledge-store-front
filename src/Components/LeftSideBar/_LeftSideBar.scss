.leftSideBar {
  height: 100%;
  width: 250px;
  position: fixed;
  z-index: 2;
  top: 56px;
  left: 0;
  background-color: #e0e0e0!important;
  overflow-x: hidden;
  // padding-top: 40px;
  -webkit-transition: 0.5s;
  -o-transition: 0.5s;
  transition: 0.5s;
  border-top: 1px solid #a7a5a5;
  font-weight: 100;

  .sidebar-navlink {
    padding-right: 25px;
    padding-top: 25px;
    display: grid;
    grid-template-columns: 20% 80%;
    grid-gap: 15px;

    #sideBarText {
      margin-top: -6px;
    }
  }

  .collapse-bar {
    position: absolute;
    bottom: 10%;
  }

/* The navigation menu links */
  a {
    padding: 8px 8px 8px 32px;
    text-decoration: none;
    font-size: 15px;
    color: #5a5a5a;
    display: block;
    transition: 0.3s;
    font-weight: 100;

    .dropdown-item {
      pointer-events: none !important;
    }
  }

  .user-account-mobile {
    padding-left: 0 !important;
    margin-top: 20px;

    a {
      pointer-events: none !important;
    }
  }

  /* When you mouse over the navigation links, change their color */
  a:hover {
    background-color: #eaeaea;
    border-left: 5px solid green;
    border-bottom-left-radius: 10px;
    // border-top-left-radius: 10px;
    cursor: pointer;
    color: green;
  }
}

/* Style page content - use this if you want to push the page content to the right when you open the side navigation */
#main {
  transition: margin-left .5s;
  padding: 20px;
  // margin-left: 250px;
  margin-top: 60px;
}

/* On smaller screens, where height is less than 450px, change the style of the sidenav (less padding and a smaller font size) */
@media screen and (max-width: 500px) {
  .leftSideBar {
    display: none;
  }

  .navbar {
    padding: .5rem 1rem !important;
  }
}

.dropdown-item.active {
  border-left: 5px solid green;
  background-color: #F5F5F5 !important;
  color: #020202;
}

.collapse-bar.active {
  border-left: 5px solid green;
  background-color: #F5F5F5 !important;
  color: #020202;
}
