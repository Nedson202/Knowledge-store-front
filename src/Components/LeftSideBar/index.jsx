import React, { Fragment, Component } from 'react';
import { NavLink, Link } from 'react-router-dom';
import ReactTooltip from 'react-tooltip';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
  SIDE_BAR_STATUS, LEFT_SIDE_BAR, CLOSED, OPEN, LEFT_SIDEBAR_NAV_LINKS,
  SIDE_NAV_WIDTH_270, SIDE_NAV_WIDTH_70, USER,
} from '../../settings';

class LeftSideBar extends Component {
  state = {
    isSideBarOpen: false,
    sideBarTextVisible: false,
  };

  componentDidMount() {
    return localStorage.sideBarStatus !== OPEN && this.toggleSidebar();
  }

  toggleSidebar = () => {
    const { isSideBarOpen } = this.state;
    let sideBarStatus;
    let sideBarWidth;
    let sideBarTextVisible;

    if (typeof window.orientation === 'undefined') {
      sideBarStatus = CLOSED;
      sideBarWidth = SIDE_NAV_WIDTH_70;
      sideBarTextVisible = false;
    }

    if (!isSideBarOpen) {
      sideBarStatus = CLOSED;
      sideBarWidth = SIDE_NAV_WIDTH_70;
      sideBarTextVisible = false;
    } else {
      sideBarStatus = OPEN;
      sideBarWidth = SIDE_NAV_WIDTH_270;
      sideBarTextVisible = true;
    }

    document.getElementById(LEFT_SIDE_BAR).style.width = sideBarWidth;
    localStorage.setItem(SIDE_BAR_STATUS, sideBarStatus);

    this.setState(prevState => ({
      isSideBarOpen: !prevState.isSideBarOpen,
      sideBarTextVisible,
    }));
  };

  renderAdminNavlinks = () => {
    const { user, isAuthenticated } = this.props;
    const { role } = user;
    const { sideBarTextVisible } = this.state;

    if (!isAuthenticated || role === USER) {
      return;
    }

    return (
      <Fragment>
        <div data-tip="Users">
          <NavLink
            className="dropdown-item sidebar-navlink"
            to="/users"
          >
            <ion-icon name="people" />
            { sideBarTextVisible && (
              <span
                className="sideBarText"
              >
                Users
              </span>
            )}
          </NavLink>
        </div>
      </Fragment>
    );
  };

  renderNavLinks = () => {
    const { isAuthenticated } = this.props;
    const { sideBarTextVisible } = this.state;

    const navLinksMap = LEFT_SIDEBAR_NAV_LINKS.map((navLink) => {
      const {
        key, label, icon, link
      } = navLink;

      return (
        <div data-tip={label} key={key}>
          <NavLink
            className={`
            dropdown-item sidebar-navlink
              ${key !== 0 && !isAuthenticated && 'blur'}
              `}
            to={link}
          >
            <ion-icon name={icon} />
            { sideBarTextVisible && (
              <span
                className="sideBarText"
              >
                {label}
              </span>
            )}
          </NavLink>
        </div>
      );
    });

    return navLinksMap;
  }

  render() {
    const { sideBarTextVisible } = this.state;

    return (
      <div id="myLeftSideBar" className="leftsidebar">
        <ReactTooltip effect="solid" place="right" />
        <div
          className="sidenav-collapse"
          onClick={this.toggleSidebar}
          data-testid="menu-button"
          role="button"
          tabIndex={0}
        >
          <div className="menu-icon">
            <ion-icon name="menu" />
          </div>
          { sideBarTextVisible && (
            <Link
              className="sideBarText"
              to="/"
            >
              Loresters Bookstore
            </Link>
          )}
        </div>
        {this.renderNavLinks()}
        {this.renderAdminNavlinks()}
      </div>
    );
  }
}

LeftSideBar.propTypes = {
  user: PropTypes.object,
  isAuthenticated: PropTypes.bool,
};

LeftSideBar.defaultProps = {
  user: {},
  isAuthenticated: false,
};

const mapStateToProps = ({ auth }) => ({
  isAuthenticated: auth.isAuthenticated,
  user: auth.user
});

export default connect(mapStateToProps)(LeftSideBar);
