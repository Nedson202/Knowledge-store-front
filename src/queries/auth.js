import { gql } from 'apollo-boost';

const addUser = gql`
  mutation($username: String!, $password: String!, $email: String!) {
    addUser(username: $username, password: $password, email: $email) {
      token
    }
  }
`;

const verifyEmail = gql`
  mutation($token: String!) {
    verifyEmail(token: $token) {
      token,
      message
    }
  }
`;

const sendVerificationEmail = gql`
  query($email: String!) {
    sendVerificationEmail(email: $email) {
      message
    }
  }
`;

const forgotPassword = gql`
  query($email: String!) {
    forgotPassword(email: $email) {
      message
    }
  }
`;

const resetPassword = gql`
  mutation($id: String!, $email: String!, $password: String!, $token: String!) {
    resetPassword(id: $id, email: $email, password: $password, token: $token) {
      message
      token
    }
  }
`;

const editProfile = gql`
  mutation($username: String, $email: String, $picture: String) {
    editProfile(username: $username, email: $email, picture: $picture) {
      token
      message
    }
  }
`;

const changePassword = gql`
  mutation($oldPassword: String!, $newPassword: String!) {
    changePassword(oldPassword: $oldPassword, newPassword: $newPassword) {
      token
      message
    }
  }
`;

const loginUser = gql`
mutation($username: String!, $password: String!) {
  loginUser(username: $username, password: $password) {
    username,
    email,
    token
  }
}
`;

const getBooks = gql`
  {
    books {
      name,
      authors,
      genre,
      year
    }
  }
`;

export {
  addUser, loginUser, getBooks, changePassword,
  verifyEmail, sendVerificationEmail,
  forgotPassword, resetPassword, editProfile
};