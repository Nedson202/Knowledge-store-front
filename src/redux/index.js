import thunk from 'redux-thunk';
import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';

import rootReducer from './reducers/rootReducer';

const enhancer = composeWithDevTools(applyMiddleware(thunk));

const store = createStore(
  rootReducer,
  enhancer
);

export default store;
