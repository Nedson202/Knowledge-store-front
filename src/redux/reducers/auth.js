import isEmpty from 'lodash/isEmpty';
import { SET_LOGGEDIN_USER } from '../constants/actionTypes';

const initialState = {
  isAuthenticated: false,
  user: {}
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_LOGGEDIN_USER: {
      return {
        ...state,
        isAuthenticated: action.user.id && !isEmpty(action.user),
        user: action.user
      };
    }

    default: return state;
  }
};