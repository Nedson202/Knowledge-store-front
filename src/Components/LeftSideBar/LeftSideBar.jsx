import React, { Component, Fragment } from 'react';
import ReactTooltip from 'react-tooltip';
import { NavLink } from 'react-router-dom';
import './_LeftSideBar.scss';

class LeftSideBar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isSideBarOpen: false,
    };
  }

  componentDidMount() {
    return localStorage.sideBarStatus !== 'open' && this.toggleSidebar();
  }

  toggleSidebar = () => {
    const { isSideBarOpen } = this.state;
    const mainContent = document.getElementById('main');

    if (isSideBarOpen) {
      localStorage.setItem('sideBarStatus', 'open');
      document.getElementById('myLeftSideBar').style.width = '250px';
      mainContent ? mainContent.style.marginLeft = '250px' : null; {/* eslint-disable-line */}
      this.toggleSidebarText('block');
    } else {
      localStorage.setItem('sideBarStatus', 'closed');
      document.getElementById('myLeftSideBar').style.width = '70px';
      mainContent ? mainContent.style.marginLeft = 'auto' : null; {/* eslint-disable-line */}
      this.toggleSidebarText('none');
    }
    this.setState(prevState => ({
      isSideBarOpen: !prevState.isSideBarOpen,
    }));
  };

  toggleSidebarText = (display) => {
    const texts = document.querySelectorAll('#sideBarText');
    for (let text = 0; text < texts.length; text += 1) {
      texts[text].style.display = display;
    }
  };

  renderAdminNavlinks = () => (
    <Fragment>
      <div data-tip="My Favorites">
        <NavLink to="/favourites" className="dropdown-item sidebar-navlink">
          <i className="fas fa-bookmark" />
          <span id="sideBarText">My Favourites</span>
        </NavLink>
      </div>
    </Fragment>
  );

  render() {
    return (
      <div id="myLeftSideBar" className="leftSideBar">
        <ReactTooltip effect="solid" place="right" />
        <div data-tip="My Books">
          <NavLink to="/my-books" className="dropdown-item sidebar-navlink">
            <i className="fa fa-book" />
            <span id="sideBarText">My Books</span>
          </NavLink>
        </div>
        <div data-tip="All Books">
          <NavLink to="/books" className="dropdown-item sidebar-navlink">
            <i className="fab fa-accusoft" />
            <span id="sideBarText">All Books</span>
          </NavLink>
        </div>
        <div data-tip="My Favorites">
          <NavLink to="/favorites" className="dropdown-item sidebar-navlink">
            <i className="fas fa-bookmark" />
            <span id="sideBarText">My Favorites</span>
          </NavLink>
        </div>
        <div data-tip="Users">
          <NavLink to="/users" className="dropdown-item sidebar-navlink">
            <i className="fa fa-users" />
            <span id="sideBarText">Users</span>
          </NavLink>
        </div>
        <div data-tip="Profile">
          <NavLink to="/profile" className="dropdown-item sidebar-navlink">
            <i className="fa fa-user-circle" />
            <span id="sideBarText">Profile</span>
          </NavLink>
        </div>
        <div data-tip="Analytics">
          <NavLink to="/analytics" className="dropdown-item sidebar-navlink">
            <i className="far fa-chart-bar" />
            <span id="sideBarText">Analytics</span>
          </NavLink>
        </div>
        <div data-tip="My Cart">
          <NavLink to="/cart" className="dropdown-item sidebar-navlink">
            <i className="fas fa-shopping-cart" />
            <span id="sideBarText">My Cart</span>
          </NavLink>
        </div>
        <div data-tip="Admins">
          <NavLink to="/admins" className="dropdown-item sidebar-navlink">
            <i className="fas fa-user-ninja" />
            <span id="sideBarText">Admins</span>
          </NavLink>
        </div>
        {this.renderAdminNavlinks()}
        <div onClick={this.toggleSidebar}> {/* eslint-disable-line */}
          <div className="dropdown-item sidebar-navlink collapse-bar" data-tip="Toggle sidebar">
            <i className="fa fa-bars" />
            <span id="sideBarText">Toggle me</span>
          </div>
        </div>
      </div>
    );
  }
}

export default LeftSideBar;
