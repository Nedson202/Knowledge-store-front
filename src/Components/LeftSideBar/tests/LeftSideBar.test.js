import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { createStore } from 'redux';
import appStore from '../../../redux';
import auth from '../../../redux/reducers/auth';
import LeftSideBar from '..';
import { LEFT_SIDEBAR_NAV_LINKS } from '../../../settings';

describe('LeftSideBar', () => {
  it('should display nav links', () => {
    const {
      container, asFragment, getByText, getByTestId
    } = render(
      <Provider store={appStore}>
        <BrowserRouter>
          <LeftSideBar />
        </BrowserRouter>
      </Provider>
    );

    const menuButton = getByTestId('menu-button');

    getByText('Loresters Bookstore');


    expect(asFragment()).toMatchSnapshot();
    expect(container).toBeInTheDocument();

    LEFT_SIDEBAR_NAV_LINKS.forEach(({ label }) => {
      getByText(label);
    });

    fireEvent.click(menuButton);
  });

  it('should render admin nav links', () => {
    const mockProps = {
      auth: {
        isAuthenticated: true,
        user: {
          role: 'admin'
        }
      }
    };

    const store = createStore(auth, mockProps);

    const { getByText } = render(
      <Provider store={store}>
        <BrowserRouter>
          <LeftSideBar />
        </BrowserRouter>
      </Provider>
    );

    getByText('Users');
  });
});
