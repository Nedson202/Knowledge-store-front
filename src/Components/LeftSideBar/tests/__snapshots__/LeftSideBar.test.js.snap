// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeftSideBar should display nav links 1`] = `
<DocumentFragment>
  <div
    class="leftsidebar"
    id="myLeftSideBar"
    style="width: 70px;"
  >
    <div
      class="__react_component_tooltip place-right type-dark "
      data-id="tooltip"
    />
    <div
      class="sidenav-collapse"
      data-testid="menu-button"
      role="button"
      tabindex="0"
    >
      <div
        class="menu-icon"
      >
        <ion-icon
          name="menu"
        />
      </div>
      <a
        class="hide sideBarText"
        href="/"
      >
        Loresters Bookstore
      </a>
    </div>
    <div
      currentitem="false"
      data-tip="All Books"
    >
      <a
        class="
              dropdown-item sidebar-navlink
              false
            "
        href="/books"
      >
        <ion-icon
          name="bookmarks"
        />
        <span
          class="hide sideBarText"
        >
          All Books
        </span>
      </a>
    </div>
    <div
      currentitem="false"
      data-tip="My Books"
    >
      <a
        class="
              dropdown-item sidebar-navlink
              blur
            "
        href="/my-books"
      >
        <ion-icon
          name="book"
        />
        <span
          class="hide sideBarText"
        >
          My Books
        </span>
      </a>
    </div>
    <div
      currentitem="false"
      data-tip="My Favorites"
    >
      <a
        class="
              dropdown-item sidebar-navlink
              blur
            "
        href="/my-favorites"
      >
        <ion-icon
          name="bookmark"
        />
        <span
          class="hide sideBarText"
        >
          My Favorites
        </span>
      </a>
    </div>
    <div
      currentitem="false"
      data-tip="My Profile"
    >
      <a
        class="
              dropdown-item sidebar-navlink
              blur
            "
        href="/my-profile"
      >
        <ion-icon
          name="person"
        />
        <span
          class="hide sideBarText"
        >
          My Profile
        </span>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
